pages/project-version/middleware/index.js

```diff
@@ -164,21 +164,57 @@ const getPreviousVersion = () => (req, res, next) => {
   }
 };
 
-const getVersionChanges = () => (req, res, next) => {
+const getGrantedVersion = () => (req, res, next) => {
+
+  if (req.project.granted) {
+
+    req.api(`/establishments/${req.establishmentId}/projects/${req.projectId}/project-versions/${req.project.granted.id}`)
+      .then(({ json: { data } }) => {
+        req.grantedVersion = data;
+      })
+      .then(() => next())
+      .catch(next);
+
+  } else {
+    next();
+  }
+};
+
+const getChanges = (current, version) => {
+  const cvKeys = traverse(current.data);
+  const pvKeys = traverse(version.data);
+  const added = remove(cvKeys, k => !pvKeys.includes(k));
+  const removed = remove(pvKeys, k => !cvKeys.includes(k));
+  let changed = [];
+  cvKeys.forEach(k => {
+    let cvNode = getNode(current.data, k);
+    let pvNode = getNode(version.data, k);
+    if (!isEqual(cvNode, pvNode)) {
+      changed.push(k);
+    }
+  });
+  return added.concat(removed).concat(changed);
+};
+
+const getLatestChanges = () => (req, res, next) => {
   if (req.prevVersion) {
-    const cvKeys = traverse(req.version.data);
-    const pvKeys = traverse(req.prevVersion.data);
-    const added = remove(cvKeys, k => !pvKeys.includes(k));
-    const removed = remove(pvKeys, k => !cvKeys.includes(k));
-    let changed = [];
-    cvKeys.forEach(k => {
-      let cvNode = getNode(req.version.data, k);
-      let pvNode = getNode(req.prevVersion.data, k);
-      if (!isEqual(cvNode, pvNode)) {
-        changed.push(k);
-      }
-    });
-    res.locals.static.changed = added.concat(removed).concat(changed);
+    res.locals.static.changes = {
+      latest: getChanges(req.version, req.prevVersion)
+    };
+  }
+  next();
+};
+
+const getGrantedChanges = () => (req, res, next) => {
+  if (req.grantedVersion) {
+    let granted = getChanges(req.version, req.grantedVersion);
+    if (res.locals.static.changes.latest) {
+      granted = granted.filter(e => !res.locals.static.changes.latest.includes(e));
+    }
+    res.locals.static.changes = {
+      ...res.locals.static.changes,
+      granted
+    };
```

Comment:
`getGrantedChanges` should not depend on `getLatestChanges` - if these are called out of turn you'll end up trying to call an array spread on `undefined` - We should make getLatestChanges and getGranted changes just plain old functions which take the request, and resolve with the required changes, we'd then just export a middleware which adds the changes to the response - something like:

```js
const getAllChanges = (req, res, next) => {
  Promise.all([
    getLatestChanges(req),
    getGrantedChanges(req)
  ])
    .then(([latest, granted]) => {
      res.locals.static.changes = { latest, granted };
    })
    .then(() => next())
    .catch(next);
}
```

Reply:
done